python/pyiceberg/expressions/base.py

```diff
@@ -342,151 +313,197 @@ def __str__(self) -> str:
+        bound_type = term.ref().field.field_type
+        if type(bound_type) in {FloatType, DoubleType}:
+            return super().__new__(cls)
+        return AlwaysFalse()
 
-class BoundIsNaN(BoundPredicate[T]):
     def __invert__(self) -> BoundNotNaN:
         return BoundNotNaN(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("IsNaN is a unary predicate and takes no Literals.")
+
+class BoundNotNaN(BoundUnaryPredicate[T]):
+    def __new__(cls, term: BoundTerm[T]):
+        bound_type = term.ref().field.field_type
+        if type(bound_type) in {FloatType, DoubleType}:
+            return super().__new__(cls)
+        return AlwaysTrue()
+
+    def __invert__(self) -> BoundIsNaN:
+        return BoundIsNaN(self.term)
 
 
-class NotNaN(UnboundPredicate[T]):
+class IsNaN(UnaryPredicate[T]):
+    as_bound = BoundIsNaN
+
+    def __invert__(self) -> NotNaN:
+        return NotNaN(self.term)
+
+
+class NotNaN(UnaryPredicate[T]):
+    as_bound = BoundNotNaN
+
     def __invert__(self) -> IsNaN:
         return IsNaN(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("NotNaN is a unary predicate and takes no Literals.")
 
-    def bind(self, schema: Schema, case_sensitive: bool) -> BoundNotNaN[T]:
-        bound_ref = self.term.bind(schema, case_sensitive)
-        return BoundNotNaN(bound_ref)
+@dataclass(frozen=True)
+class SetPredicate(Unbound[T, BooleanExpression], BooleanExpression, ABC):
+    as_bound: ClassVar[type]
+    term: UnboundTerm[T]
+    literals: tuple[Literal[T], ...]
+
+    @abstractmethod
+    def __invert__(self) -> BooleanExpression:
+        ...
 
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
+        bound_term = self.term.bind(schema, case_sensitive)
+        return self.as_bound(bound_term, {lit.to(bound_term.ref().field.field_type) for lit in self.literals})
 
-class BoundNotNaN(BoundPredicate[T]):
-    def __invert__(self) -> BoundIsNaN:
-        return BoundIsNaN(self.term)
 
-    def _validate_literals(self):  # pylint: disable=W0238
-        if self.literals:
-            raise AttributeError("NotNaN is a unary predicate and takes no Literals.")
+@dataclass(frozen=True)
+class BoundSetPredicate(Bound[T], BooleanExpression, ABC):
+    term: BoundTerm[T]
+    literals: set[Literal[T]]
 
+    @abstractmethod
+    def __invert__(self) -> BooleanExpression:
+        ...
 
-class BoundIn(BoundPredicate[T]):
-    def _validate_literals(self):  # pylint: disable=W0238
-        if not self.literals:
-            raise AttributeError("BoundIn must contain at least 1 literal.")
 
-    def __invert__(self) -> BoundNotIn[T]:
-        return BoundNotIn(self.term, *self.literals)
+class BoundIn(BoundSetPredicate[T]):
+    def __new__(cls, term: BoundTerm[T], literals: set[Literal[T]]) -> BooleanExpression:
```

Comment:
We could also consider wrapping it into positional arguments:
```python
class BoundIn(BoundSetPredicate[T]):
    def __new__(cls, term: BoundTerm[T], *literals: Literal[T]) -> BooleanExpression:
        literals_set = set(literals)
        count = len(literals)
        if count == 0:
            return AlwaysFalse()
        if count == 1:
            return BoundEq(term, literals.pop())
        else:
            return super().__new__(cls)

    def __invert__(self) -> BooleanExpression:
        return BoundNotIn(self.term, self.literals)
```
This way we also don't have to care about mutating the set. See comment below 👍🏻 

This way we don't have to pass in a set:
```python
BoundIn[str](
    base.BoundReference(
        field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
        accessor=Accessor(position=0, inner=None),
    ),
    StringLiteral("foo"), StringLiteral("bar"), StringLiteral("baz"),
)
```
Instead of:
```python
BoundIn[str](
    base.BoundReference(
        field=NestedField(field_id=1, name="foo", field_type=StringType(), required=False),
        accessor=Accessor(position=0, inner=None),
    ),
    {StringLiteral("foo"), StringLiteral("bar"), StringLiteral("baz")},
)
```